File: backend/app/schemas/documents.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.documents import (
    DocumentClassification,
    DocumentType,
    DocumentVisibility,
    IngestionStatus,
)


class DocumentUploadMetadata(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)

    tenant_id: UUID
    company_id: UUID
    department: str = Field(pattern=r"^(finance|legal|shared)$")
    visibility: DocumentVisibility
    classification: DocumentClassification
    document_type: DocumentType
    reporting_period: str | None = Field(
        default=None,
        min_length=1,
        max_length=64,
        pattern=r"^[A-Za-z0-9][A-Za-z0-9 ._/-]*$",
    )

    @model_validator(mode="after")
    def validate_classification(self) -> "DocumentUploadMetadata":
        required_pair = {
            "finance": (
                DocumentVisibility.DEPARTMENT_PRIVATE,
                DocumentClassification.FINANCE_ONLY,
            ),
            "legal": (
                DocumentVisibility.DEPARTMENT_PRIVATE,
                DocumentClassification.LEGAL_ONLY_CONFIDENTIAL,
            ),
            "shared": (
                DocumentVisibility.TENANT_SHARED,
                DocumentClassification.TENANT_SHARED,
            ),
        }[self.department]
        if (self.visibility, self.classification) != required_pair:
            raise ValueError("Department, visibility, and classification are inconsistent")
        if (
            self.document_type in {DocumentType.FINANCIAL_REPORT, DocumentType.SPREADSHEET}
            and self.reporting_period is None
        ):
            raise ValueError("Reporting period is required for financial reports and spreadsheets")
        return self


class WarningData(BaseModel):
    code: str
    message: str


class DocumentScopeData(BaseModel):
    tenant_id: UUID
    tenant_slug: str
    tenant_name: str
    company_id: UUID
    company_slug: str
    company_name: str
    department: str
    visibility: DocumentVisibility
    classification: DocumentClassification


class DocumentVersionData(BaseModel):
    id: UUID
    version_number: int
    original_filename: str
    media_type: str
    source_type: Literal["PDF", "XLSX", "CSV", "UNKNOWN"]
    checksum_sha256: str
    size_bytes: int
    status: IngestionStatus
    page_count: int
    sheet_count: int
    row_count: int
    cell_count: int
    warnings: tuple[WarningData, ...]
    uploaded_by_user_id: UUID
    approved_by_user_id: UUID | None
    created_at: datetime


class DocumentData(BaseModel):
    id: UUID
    scope: DocumentScopeData
    document_type: DocumentType
    reporting_period: str | None
    current_approved_version_id: UUID | None
    version: DocumentVersionData
    ingestion_job_id: UUID


class UploadResultData(BaseModel):
    document: DocumentData
    deduplicated: bool


class IngestionStatusData(BaseModel):
    ingestion_job_id: UUID
    document_id: UUID
    document_version_id: UUID
    version_number: int
    status: IngestionStatus
    safe_error_code: str | None
    warnings: tuple[WarningData, ...]
    page_count: int
    sheet_count: int
    row_count: int
    cell_count: int
    started_at: datetime | None
    completed_at: datetime | None
    updated_at: datetime


class ParsedPageData(BaseModel):
    page_number: int
    text: str


class ParsedCellData(BaseModel):
    row_number: int
    column_number: int
    coordinate: str
    value: str
    value_kind: str
    formula_like: bool


class ParsedRowData(BaseModel):
    row_number: int
    cells: tuple[ParsedCellData, ...]


class ParsedSheetData(BaseModel):
    sheet_index: int
    name: str
    row_count: int
    column_count: int
    rows: tuple[ParsedRowData, ...]


class DocumentPreviewData(BaseModel):
    document: DocumentData
    pages: tuple[ParsedPageData, ...]
    sheets: tuple[ParsedSheetData, ...]


class VersionActionData(BaseModel):
    document_id: UUID
    document_version_id: UUID
    status: IngestionStatus
    current_approved_version_id: UUID | None


class DeleteDocumentData(BaseModel):
    document_id: UUID
    status: IngestionStatus = IngestionStatus.DELETED


class DocumentListData(BaseModel):
    items: tuple[DocumentData, ...]
    total: int
    limit: int
    offset: int


class CompanyOptionData(BaseModel):
    id: UUID
    slug: str
    name: str


class TenantOptionData(BaseModel):
    id: UUID
    slug: str
    name: str
    companies: tuple[CompanyOptionData, ...]


class ClassificationPairData(BaseModel):
    department: str
    visibility: DocumentVisibility
    classification: DocumentClassification
    label: str


class DocumentTypeOptionData(BaseModel):
    value: DocumentType
    label: str
    reporting_period_required: bool


class UploadLimitsData(BaseModel):
    max_upload_bytes: int
    extensions: tuple[str, ...]
    mime_types: tuple[str, ...]


class DocumentOptionsData(BaseModel):
    tenants: tuple[TenantOptionData, ...]
    classification_pairs: tuple[ClassificationPairData, ...]
    document_types: tuple[DocumentTypeOptionData, ...]
    limits: UploadLimitsData

File: frontend/src/types/documents.ts
export type IngestionStatus =
  | 'UPLOADED'
  | 'VALIDATING'
  | 'PARSING'
  | 'PREVIEW_READY'
  | 'APPROVED'
  | 'REJECTED'
  | 'VALIDATION_FAILED'
  | 'PARSING_FAILED'
  | 'DELETED'

export type DocumentVisibility = 'DEPARTMENT_PRIVATE' | 'TENANT_SHARED'

export type DocumentClassification =
  'FINANCE_ONLY' | 'LEGAL_ONLY_CONFIDENTIAL' | 'TENANT_SHARED'

export type DocumentType =
  | 'FINANCIAL_REPORT'
  | 'LEGAL_AGREEMENT'
  | 'POLICY'
  | 'EMAIL'
  | 'SPREADSHEET'
  | 'OTHER'

export type DocumentSourceType = 'PDF' | 'XLSX' | 'CSV' | 'UNKNOWN'

export interface CompanyOption {
  id: string
  slug: string
  name: string
}

export interface TenantOption {
  id: string
  slug: string
  name: string
  companies: CompanyOption[]
}

export interface ClassificationPairOption {
  department: string
  visibility: DocumentVisibility
  classification: DocumentClassification
  label: string
}

export interface DocumentTypeOption {
  value: DocumentType
  label: string
  reporting_period_required: boolean
}

export interface IngestionOptionsData {
  tenants: TenantOption[]
  classification_pairs: ClassificationPairOption[]
  document_types: DocumentTypeOption[]
  limits: {
    max_upload_bytes: number
    extensions: string[]
    mime_types: string[]
  }
}

export interface UploadMetadata {
  tenant_id: string
  company_id: string
  department: string
  visibility: DocumentVisibility
  classification: DocumentClassification
  document_type: DocumentType
  reporting_period: string | null
}

export interface WarningData {
  code: string
  message: string
}

export interface ApiDocumentScopeData {
  tenant_id: string
  tenant_slug: string
  tenant_name: string
  company_id: string
  company_slug: string
  company_name: string
  department: string
  visibility: DocumentVisibility
  classification: DocumentClassification
}

export interface ApiDocumentVersionData {
  id: string
  version_number: number
  original_filename: string
  media_type: string
  source_type: DocumentSourceType
  checksum_sha256: string
  size_bytes: number
  status: IngestionStatus
  page_count: number
  sheet_count: number
  row_count: number
  cell_count: number
  warnings: WarningData[]
  uploaded_by_user_id: string
  approved_by_user_id: string | null
  created_at: string
}

export interface ApiDocumentData {
  id: string
  scope: ApiDocumentScopeData
  document_type: DocumentType
  reporting_period: string | null
  current_approved_version_id: string | null
  version: ApiDocumentVersionData
  ingestion_job_id: string
}

export interface ApiUploadResultData {
  document: ApiDocumentData
  deduplicated: boolean
}

export interface ApiIngestionStatusData {
  ingestion_job_id: string
  document_id: string
  document_version_id: string
  version_number: number
  status: IngestionStatus
  safe_error_code: string | null
  warnings: WarningData[]
  page_count: number
  sheet_count: number
  row_count: number
  cell_count: number
  started_at: string | null
  completed_at: string | null
  updated_at: string
}

export interface ApiParsedPageData {
  page_number: number
  text: string
}

export interface ApiParsedCellData {
  row_number: number
  column_number: number
  coordinate: string
  value: string
  value_kind: string
  formula_like: boolean
}

export interface ApiParsedRowData {
  row_number: number
  cells: ApiParsedCellData[]
}

export interface ApiParsedSheetData {
  sheet_index: number
  name: string
  row_count: number
  column_count: number
  rows: ApiParsedRowData[]
}

export interface ApiDocumentPreviewData {
  document: ApiDocumentData
  pages: ApiParsedPageData[]
  sheets: ApiParsedSheetData[]
}

export interface ApiVersionActionData {
  document_id: string
  document_version_id: string
  status: IngestionStatus
  current_approved_version_id: string | null
}

export interface ApiDocumentListData {
  items: ApiDocumentData[]
  total: number
  limit: number
  offset: number
}

export interface IngestionJobData {
  job_id: string
  document_id: string
  version_id: string
  version_number: number
  status: IngestionStatus
  warnings: string[]
  safe_error_code: string | null
  page_count: number
  sheet_count: number
  row_count: number
  cell_count: number
  updated_at: string
}

export interface UploadDocumentData extends IngestionJobData {
  filename: string
  checksum: string
  deduplicated: boolean
}

export interface DocumentSummary {
  document_id: string
  version_id: string
  version_number: number
  filename: string
  checksum: string
  source_type: DocumentSourceType
  detected_mime_type: string
  size_bytes: number
  tenant: Omit<TenantOption, 'companies'>
  company: CompanyOption
  department: string
  visibility: DocumentVisibility
  classification: DocumentClassification
  document_type: DocumentType
  reporting_period: string | null
  status: IngestionStatus
  page_count: number
  sheet_count: number
  row_count: number
  cell_count: number
  uploader_id: string
  approved_by_user_id: string | null
  current_approved_version_id: string | null
  warnings: string[]
  created_at: string
  ingestion_job_id: string
}

export interface DocumentLibraryData {
  items: DocumentSummary[]
  total: number
  limit: number
  offset: number
}

export interface ParsedPageData {
  page_number: number
  text: string
}

export interface ParsedCellData {
  coordinate: string
  value: string
  value_kind: string
  formula_like: boolean
}

export interface ParsedRowData {
  row_number: number
  cells: ParsedCellData[]
}

export interface ParsedSheetData {
  sheet_index: number
  sheet_name: string
  row_count: number
  column_count: number
  rows: ParsedRowData[]
}

export interface PdfPreviewContent {
  kind: 'pdf'
  page_count: number
  pages: ParsedPageData[]
}

export interface SpreadsheetPreviewContent {
  kind: 'spreadsheet'
  source_type: 'XLSX' | 'CSV'
  sheet_count: number
  sheets: ParsedSheetData[]
}

export interface DocumentPreviewData {
  document: DocumentSummary
  warnings: string[]
  content: PdfPreviewContent | SpreadsheetPreviewContent
}

export interface DocumentMutationData {
  document_id: string
  version_id: string
  status: IngestionStatus
  current_approved_version_id: string | null
}

export interface DeleteDocumentData {
  document_id: string
  status: 'DELETED'
}

export interface DocumentFilters {
  tenant_id?: string
  company_id?: string
  department?: string
  document_type?: DocumentType
  status?: IngestionStatus
  offset?: number
  limit?: number
}
